Add support for bokeh event callbacks

philippjfr opened the pull request:
Adds support for the upcoming event callbacks in bokeh, currently being developed [here](https://github.com/jlstevens/bokeh/pull/4).

Summary of how it works
===================

1. Check if any of the ``skip_conditions`` apply, if so, skip

2. Grab all the requested ``attributes`` from the models and accumulate them in data

3. Check if there is a ``Comm`` instantiated otherwise open it and store it in a global registry

4. Check if there is a ``comm_state`` for the ``Comm`` otherwise make one defining a ``event_buffer``, a ``blocked`` flag and a ``time``

5. Check whether the comm_state is blocked or has timed out (i.e. ``Date.now()>comm_state.time+timeout``):

   a) if it's blocked put data along with ``cb_obj.event.event_name`` on top of the ``event_buffer``

   b) if it's not blocked also prepend to ``event_buffer`` but also call ``process_event`` with small timeout. The small timeout acts as a debouncing mechanism as new events can be pushed ahead of the original event in the event queue. After process_events is called set the ``comm_state`` to blocked

6. ``process_events`` will iterate over the ``event_buffer`` and filter out older duplicated events. This ensures only the latest event of each type is processed (e.g. to ensure ``panstart``, ``pan`` and ``panend`` are all processed). (Can probably be simplified)

7. In python everything does its thing and sends back a message either acknowledging execution or failure. This message also contains the ``comm_id``, with which the ``comm_state`` is looked up. If there is something on the ``event_buffer`` process it and keep the ``Comm`` blocked, otherwise unblock.

ToDo:

- [x] Update docstring
- [x] Finalize naming for Callback attributes
- [ ] Add tests to ensure callbacks reference correct handles/models
- [ ] Add documentation about defining custom Stream/Callbacks

jlstevens commented on holoviews/plotting/bokeh/callbacks.py:
```diff
@@ -92,19 +104,41 @@ class Callback(object):
     attributes = {}
 
     js_callback = """
+        function unique_events(events) {{
```
As far as I can tell this is a self-contained function that could be a utility. Would be good to have a docstring i.e a comment line saying that it does and the same for ``process_events`` below. At some point we will probably want the concept of Javascript 'utilities' that you can just include where needed.

That said, I just noticed ``comm_state`` is referenced in this function, can it not be made into an explicit function argument?

philippjfr replied:
This is not ready to review. I've got outstanding changes locally.

philippjfr replied:
Independent of the details of this function a set of JS utilities definitely makes sense. The issue then is that we will really want to have a proper JS package, because polluting the main namespace isn't ideal. All things to consider once we get started writing the widget manager.

jlstevens replied:
Agreed.

jlstevens replied:
Ok, the signature now looks fine and making JS utilities is for some other day.

jlstevens commented on holoviews/plotting/bokeh/callbacks.py:
```diff
@@ -92,19 +104,41 @@ class Callback(object):
+        }}
+
+        function process_events(comm_state) {{
+            var events = unique_events(comm_state.events);
```
So this is at least one place where ``unique_events`` is used and ``comm_state`` is available. You could definitely pass ``comm_state`` to ``unique_events`` here though ``unique_events(comm_state.events, comm_state)`` is a bit redundant (just pass ``comm_state``). The fact that you can access ``comm_state`` at all in ``unique_event`` without explicitly declaring it makes me uncomfortable (but then again, this is Javascript!)

philippjfr replied:
As I said above, I would not review this yet.

jlstevens replied:
Fair enough - I was just trying to understand the code out of my own curiosity and thought I might as well make some comments while I'm at it.

philippjfr replied:
Looking is fine, but there's no point reviewing things I'm still actively working on.

jlstevens commented on holoviews/plotting/bokeh/element.py:
```diff
@@ -104,6 +103,12 @@ class ElementPlot(BokehPlot, GenericElementPlot):
+            'fixed', 'scale_width', 'scale_height',
+            'scale_both', 'stretch_both'
+        ], doc="""Defines how the plot scales when resized.""")
+
```
Is this relevant to hooking up to events?

philippjfr replied:
Please stop reviewing this PR.

jlstevens replied:
Why not just work on a branch if it is totally WIP? Once you make a PR you make the proposed changes public in a way that invites comment even if the status is WIP.

I would recommend working on a branch and only turning it into a PR when you require some sort of feedback.

philippjfr replied:
>I would recommend working on a branch and only turning it into a PR when you require some sort of feedback.

I disagree, I like being able to easily see an overview of my changes on Github and that's exactly what the WIP and ready tags are for.

jbednar replied:
But if it's really not ready for someone to look at, in addition to tagging it WIP you should put a sentence in the description saying what you're up to so that other people know what the status us.  Lots of things marked WIP are ready to look at, just not ready to merge.

jlstevens replied:
A WIP tag says that the *pull request* is in work in progress which means there is a public conversation going on about a feature that is in progress. That isn't the same as 'please do not comment' and it sounds like that is how you want to define it.

Github defines a [PR](https://help.github.com/articles/about-pull-requests/) as:

> Pull requests let you tell others about changes you've pushed to a repository on GitHub.

It is clearly there for the purposes of communication, even if you are communicating to yourself (e.g to see an overview of your changes). I'm wondering if there is a tag that better conveys what you intended - not only 'work in progress' but 'don't supply feedback yet'.

philippjfr replied:
Agreed, I'll add comments, but the usage of WIP across all the projects I've encountered has meant not ready for review, and ``ready`` means ready for review not necessarily merge.

philippjfr replied:
A PR is a place for discussion, which is often helpful long before the code is ready to review.

jlstevens replied:
Ah, I see Jim agrees with me here. :-)

'Not ready for review' is a bit long, but could just about be a tag? I also agree that there are different conventions across projects which is why a WIP tag by itself is ambiguous.

jbednar replied:
What it usually means in the projects I'm familiar with is *definitely* not ready to merge, *probably* not ready for heavyhanded, very detailed comments, and *probably* ready for someone to take a quick look and start giving some feedback.  If that's not the case for a given WIP PR, just declare it in text so that other humans know what to do with it.  No big deal...

jlstevens replied:
How about a ``pre-review`` tag if you don't want comment? I'm not sure it replaces the ``WIP`` tag but from the sound of it, it might do.

philippjfr replied:
I think high-level feedback on a WIP PR makes sense but not necessarily which arguments should be passed to which functions. I think we should just clarify what the tags mean.

jlstevens replied:
The problem with a PR like this is that the description at the top is very brief and nothing worth commenting on. That means to know what the PR is really about I have to look at the code, at which point it becomes very tempting to make comments, even without delving into the exact details i.e just trying to understand what is going on.

That is probably why what Jim suggests makes the most sense - just say you aren't ready to receive comments just yet at the top.

philippjfr replied:
>That is probably why what Jim suggests makes the most sense - just say you aren't ready to receive comments just yet at the top.

Agreed, this PR has been around a while and I forgot to update the description. That said I'm still pretty firm on the meanings of ``WIP`` and ``ready``, but without putting that information somewhere I guess no one will know what those conventions are.

jlstevens replied:
How about ``WIP`` and ``WIP-discussion``? Then we can have the semantics you want.

jbednar replied:
Or `WIP` and `WIP-pre-review`...

jlstevens replied:
Then ``WIP`` allows for discussion. :-)

That said, I *believe* you can rename GitHub tags so all existing PRs and issues get renamed at once. Personally, I like Jim's suggestion more but I am happy to be overruled.

jbednar replied:
Also, I personally prefer that a WIP PR includes an explicit checklist in the description, stating what the author intends to do before it will be ready, and marking those off as they are done.  That way it's fully ready for review about *some* things, and other things are clearly not ready for review.  This PR doesn't have such a checklist, which is one reason it's hard for other people to know what to do with it.

jlstevens replied:
By the way, GitHub supports PR templates which is where this information could go. I would keep it short as a long list of requirements/restrictions/demands would put people off contributing.

Asking for a checklist of items would be a very reasonable thing to do though.

philippjfr replied:
I agree, I definitely think we should add a PR and Issue template.

philippjfr commented:
1. Check if any of the ``skip_conditions`` apply, if so, skip
2. Grab all the requested ``attributes`` from the models and accumulate them in data
3. Check if there is a ``Comm`` instantiated otherwise open it and store it in a global registry
4. Check if there is a ``comm_state`` for the ``Comm`` otherwise make one defining a ``event_buffer``, a ``blocked`` flag and a ``time``
5. Check whether the comm_state is blocked or has timed out (i.e. ``Date.now()>comm_state.time+timeout``):
a) if it's blocked put data along with ``cb_obj.event.event_name`` on top of the ``event_buffer``
b) if it's not blocked also prepend to ``event_buffer`` but also call ``process_event`` with small timeout. The small timeout acts as a debouncing mechanism as new events can be pushed ahead of the original event in the event queue. After process_events is called set the ``comm_state`` to blocked
``process_events`` will iterate over the ``event_buffer`` and filter out older duplicated events. This ensures only the latest event of each type is processed (e.g. to ensure ``panstart``, ``pan`` and ``panend`` are all processed). (Can probably be simplified)

6. In python everything does its thing and sends back a message either acknowledging execution or failure. This message also contains the ``comm_id``, with which the ``comm_state`` is looked up. If there is something on the ``event_buffer`` process it and keep the ``Comm`` blocked, otherwise unblock.

philippjfr commented:
Ready for review.

jlstevens commented on holoviews/plotting/bokeh/element.py:
```diff
@@ -370,9 +374,11 @@ def _init_plot(self, key, element, plots, ranges=None):
-        return bokeh.plotting.Figure(x_axis_type=x_axis_type,
-                                     y_axis_type=y_axis_type, title=title,
-                                     **properties)
+        with warnings.catch_warnings():
```
Probably worth mentioning what warning you are catching and why it doesn't matter that you are suppressing it.

jlstevens commented on holoviews/plotting/bokeh/callbacks.py:
```diff
@@ -53,35 +55,55 @@ def attributes_js(attributes, handles):
-    * attributes :  The attributes define which attributes to send
+    * extra_models: Any additional models available in handles which
+                    should be made available in the namespace of the
+                    objects.
```
Giving an example of why you might want this would help here. E.g 'For instance, this can be used to check which tools are currently active'.

jlstevens commented on holoviews/plotting/bokeh/callbacks.py:
```diff
@@ -53,35 +55,55 @@ def attributes_js(attributes, handles):
+                    specified as a list of valid JS expressions, which
+                    can reference models requested by the callback,
+                    e.g. ['pan.attributes.active'] would skip the
+                    callback if the pan tool is active.
```
If the models references in the skip conditions need to be in ``extra_models`` it would be helpful to say so.

philippjfr replied:
That's what this line was about:

>which can reference models requested by the callback,

but I'll try to clarify.

jlstevens commented on holoviews/plotting/bokeh/callbacks.py:
```diff
@@ -129,35 +178,42 @@ class Callback(object):
-
-        timeout = comm_state.timeout + {timeout};
+        // Add current event to queue and process queue if not blocked
+        event_name = cb_obj.event ? cb_obj.event.event_name : undefined
```
I think this line is expected to change - I've been requested to change how this works in Bokeh.

philippjfr replied:
Okay, can fix that in a follow-up PR when you've made that change.

jlstevens commented on holoviews/plotting/bokeh/callbacks.py:
```diff
@@ -215,13 +279,18 @@ def _filter_msg(self, msg, ids):
+                                for h, hid in metadata.items()}
         Stream.trigger(self.streams)
+        for stream in self.streams:
+            stream._metadata = None
```
I would consider just declaring ``_metadata`` to be an empty dict by default. I'm assuming this loop is to reset...

jlstevens commented on holoviews/streams.py:
```diff
@@ -124,6 +129,8 @@ def __init__(self, preprocessors=[], source=None, subscribers=[], **params):
         self.subscribers = subscribers
         self.preprocessors = preprocessors
         self._hidden_subscribers = []
+        self._metadata = None
```
Should probably be an empty dict and I would add a comment what it is for - right now it is about stopping loops.

philippjfr replied:
Agreed, definitely needs a comment.

jlstevens commented on holoviews/plotting/bokeh/callbacks.py:
```diff
@@ -272,54 +340,109 @@ def set_customjs(self, handle, references):
+        data = "var data = {};\n"
+        code = conditional + data + attributes + self.code + self_callback
+
+        js_callback = CustomJS(args=references, code=code)
```
I would consider factoring out a method that returns a ``CustomJS`` so all the Javascript is kept separate (probably including the call to ``attributes_js``). The method could be called something like ``get_customjs``...

jlstevens commented on holoviews/plotting/bokeh/callbacks.py:
```diff
@@ -272,54 +340,109 @@ def set_customjs(self, handle, references):
+
+        js_callback = CustomJS(args=references, code=code)
+        cb = None
+        if id(handle) in self._callbacks:
```
Use of ``id`` here probably warrants a comment about what it is for...

jlstevens commented on holoviews/streams.py:
```diff
@@ -72,6 +72,10 @@ class Stream(param.Parameterized):
+
+    Depending on the plotting backend certain streams may interactively
+    subscribe to events and changes by the plotting backend. To disable
+    this behavior instantiate the Stream with interactive=False.
```
I wish there were a better name for this concept by ``interactive`` is the best I can think of right now...

jlstevens replied:
How about ``live``?

jlstevens replied:
Then we can pretend we are doing 'live streams' ;-p

philippjfr replied:
Prefer ``interactive`` tbh.

jlstevens replied:
@jbednar Any thoughts?

The concept here is you might want a ``PositionXY`` stream you want to manage yourself and you don't want the bokeh backend to try to supply you values. By default Bokeh supplies values but this option allows you to switch it off if you want (e.g you want to use streams in the same way between bokeh an matplotlib).

jbednar replied:
automatic? autoupdate? tied?

philippjfr replied:
``linked``?

jbednar replied:
Sure, linked.

jlstevens replied:
``linked`` isn't bad

jlstevens commented on holoviews/plotting/bokeh/callbacks.py:
```diff
@@ -215,13 +279,18 @@ def _filter_msg(self, msg, ids):
     def on_msg(self, msg):
         for stream in self.streams:
-            ids = self.handle_ids[stream]
+            metadata = self.handle_ids[stream]
```
I think I would just call this ``handle_ids`` here.

jlstevens commented on holoviews/plotting/bokeh/callbacks.py:
```diff
@@ -215,13 +279,18 @@ def _filter_msg(self, msg, ids):
                 continue
             stream.update(trigger=False, **processed_msg)
+            stream._metadata = {h: {'id': hid, 'events': self.events}
+                                for h, hid in metadata.items()}
```
... and use ``... for h, hid in handle_ids.items()`` here...

jlstevens replied:
Also, if you expect to use ``_metadata`` for anything else, then you'll need a key e.g ``'prevent_loop_ids'`` (or similar).

jlstevens commented:
There are a few minor outstanding things - just tell me when you think it is ready to merge and I'll merge (as long as the tests are passing).

philippjfr commented:
I've addressed all your comments except for the change you are still going to make in bokeh. It's just that you commented on the line above in some cases.

jlstevens commented:
Ok, merging.
